src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -32,13 +40,7 @@ public ActivitySource(string name, string? version = "")
 
             if (s_allListeners.Count > 0)
             {
-                s_allListeners.EnumWithAction(listener => {
-                    var shouldListenTo = listener.ShouldListenTo;
-                    if (shouldListenTo != null && shouldListenTo(this))
-                    {
-                        AddListener(listener);
-                    }
-                });
+                s_allListeners.EnumWithAction(ListenerShouldListenToSource, this);
```

Comment:
You don't need to manually cache these.  By keeping the callback code with the call site, it can be a lot easier to read, and Roslyn can cache the delegates for you, e.g.
```C#
s_allListeners.EnumWithAction((listener, obj) =>
{
    ActivitySource source = (ActivitySource)obj;
    var shouldListenTo = listener.ShouldListenTo;
    if (shouldListenTo != null && shouldListenTo(source))
    {
        source.AddListener(listener);
    }
}, this);
```
You just need to take care not to accidentally access any state that would cause the closure to happen, but in the very near future we should be getting static lambdas which will let the compiler flag such violations for us.

Reply:
Thanks, @stephentoub. I have reverted back using anonymous methods.